### ISSUE TITLE
OpenWRT support for Mikrotik RB493GAH

### ISSUE THREAD

**Issue by AlverGant** [openwrt/openwrt#523]
Hi,

I'm not sure if RB493GAH is supported because I am experiencing difficulties regarding NAND flash partitioning so I'm asking it here.

I am able to TFTP boot a Chaos Calmer compiled initramfs.elf binary but when I try to erase kernel and rootfs partitions, thousands of bad erase block appears leaving no room to permanent installation.
My flash chip is Samsumg K9F1208U0C which is a 64MB 8bit device, different from the most common 128MB versions.

Here is my proc/mtd parts which I do think is wrong
dev:    size   erasesize  name
mtd0: 0000b000 00001000 "routerboot"
mtd1: 00001000 00001000 "hard_config"
mtd2: 00002000 00001000 "bios"
mtd3: 00001000 00001000 "soft_config"
mtd4: 00040000 00004000 "booter"
mtd5: 003c0000 00004000 "kernel"
mtd6: 03c00000 00004000 "rootfs"

Unfortunately, now I don't have space available even to do a rollback to RouterOS via netinstall, leaving a significantly powerfull device almost useless.

If somebody can give me a hint that would be awesome.

Bellow the console output during TFTP boot

RouterBOOT booter 3.24

RouterBoard 493G

CPU frequency: 680 MHz
  Memory size: 256 MiB
    NAND size:  64 MiB

Press any key within 2 seconds to enter setup..
trying dhcp protocol... OK
resolved mac address EC:08:6B:EC:3F:A2
Gateway: 10.61.34.254
transfer started .............................................. transfer ok, time=2.75s

loading kernel from nand... kernel not found
trying dhcp protocol... OK
resolved mac address EC:08:6B:EC:3F:A2
Gateway: 10.61.34.254
transfer started .............................................. transfer ok, time=2.74s
setting up elf image... OK
jumping to kernel code
[    0.000000] Linux version 3.18.45 (lede@lede) (gcc version 4.8.3 (OpenWrt/Linaro GCC 4.8-2014.04 r49389) ) #4 Sun Aug 13 23:50:06 BRT 2017
[    0.000000] bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019374 (MIPS 24Kc)
[    0.000000] SoC: Atheros AR7161 rev 2
[    0.000000] Determined physical RAM map:
[    0.000000]  memory: 08000000 @ 00000000 (usable)
[    0.000000] User-defined physical RAM map:
[    0.000000]  memory: 10000000 @ 00000000 (usable)
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x00000000-0x0fffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x00000000-0x0fffffff]
[    0.000000] Initmem setup node 0 [mem 0x00000000-0x0fffffff]
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, VIPT, cache aliases, linesize 32 bytes
[    0.000000] Built 1 zonelists in Zone order, mobility grouping on.  Total pages: 65024
[    0.000000] Kernel command line:  console=ttyS0,115200 parts=1 boot_part_size=4194304 gpio=4023 HZ=340000000 mem=256M kmac=D4:CA:6D:34:0A:29 board=493G ver=3.24 hw_opt=00000000 boot=1 mlc=9 rootfstype=yaffs noinitrd
[    0.000000] PID hash table entries: 1024 (order: 0, 4096 bytes)
[    0.000000] Dentry cache hash table entries: 32768 (order: 5, 131072 bytes)
[    0.000000] Inode-cache hash table entries: 16384 (order: 4, 65536 bytes)
[    0.000000] Writing ErrCtl register=00000000
[    0.000000] Readback ErrCtl register=00000000
[    0.000000] Memory: 253312K/262144K available (2436K kernel code, 153K rwdata, 516K rodata, 2836K init, 184K bss, 8832K reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS:51
[    0.000000] Clocks: CPU:680.000MHz, DDR:340.000MHz, AHB:170.000MHz, Ref:40.000MHz
[    0.000000] Calibrating delay loop... 452.19 BogoMIPS (lpj=2260992)
[    0.060000] pid_max: default: 32768 minimum: 301
[    0.060000] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.070000] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.070000] NET: Registered protocol family 16
[    0.080000] MIPS: machine is MikroTik RouterBOARD 493G
[    2.710000] registering PCI controller with io_map_base unset
[    2.720000] rb4xx-spi rb4xx-spi: master is unqueued, this is deprecated
[    2.730000] PCI host bridge to bus 0000:00
[    2.730000] pci_bus 0000:00: root bus resource [mem 0x10000000-0x16ffffff]
[    2.740000] pci_bus 0000:00: root bus resource [io  0x0000]
[    2.740000] pci_bus 0000:00: No busn resource found for root bus, will use [bus 00-ff]
[    2.750000] pci 0000:00:13.0: BAR 0: assigned [mem 0x10000000-0x1000ffff]
[    2.750000] pci 0000:00:13.0: using irq 41 for pin 1
[    2.760000] Switched to clocksource MIPS
[    2.760000] NET: Registered protocol family 2
[    2.770000] TCP established hash table entries: 2048 (order: 1, 8192 bytes)
[    2.770000] TCP bind hash table entries: 2048 (order: 1, 8192 bytes)
[    2.780000] TCP: Hash tables configured (established 2048 bind 2048)
[    2.790000] TCP: reno registered
[    2.790000] UDP hash table entries: 256 (order: 0, 4096 bytes)
[    2.790000] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes)
[    2.800000] NET: Registered protocol family 1
[    5.790000] futex hash table entries: 256 (order: -1, 3072 bytes)
[    5.800000] msgmni has been set to 494
[    5.810000] io scheduler noop registered
[    5.820000] io scheduler deadline registered (default)
[    5.820000] Serial: 8250/16550 driver, 1 ports, IRQ sharing disabled
[    5.830000] console [ttyS0] disabled
[    5.850000] serial8250.0: ttyS0 at MMIO 0x18020000 (irq = 11, base_baud = 10625000) is a 16550A
[    5.860000] console [ttyS0] enabled
[    5.860000] console [ttyS0] enabled
[    5.870000] bootconsole [early0] disabled
[    5.870000] bootconsole [early0] disabled
[    5.880000] m25p80 spi0.0: pm25lv512 (64 Kbytes)
[    5.890000] Creating 4 MTD partitions on "spi0.0":
[    5.890000] 0x000000000000-0x00000000b000 : "routerboot"
[    5.900000] 0x00000000b000-0x00000000c000 : "hard_config"
[    5.900000] 0x00000000d000-0x00000000f000 : "bios"
[    5.910000] 0x00000000f000-0x000000010000 : "soft_config"
[    5.910000] NAND flash driver for RouterBoard 4xx series version 0.2.0
[    5.920000] nand: device found, Manufacturer ID: 0xec, Chip ID: 0x76
[    5.930000] nand: Samsung NAND 64MiB 3,3V 8-bit
[    5.930000] nand: 64MiB, SLC, page size: 512, OOB size: 16
[    5.940000] Scanning device for bad blocks
[    5.990000] Bad eraseblock 257 at 0x000000404000
[    5.990000] Bad eraseblock 258 at 0x000000408000
[    6.000000] Bad eraseblock 259 at 0x00000040c000
[    6.000000] Bad eraseblock 260 at 0x000000410000
[    6.010000] Bad eraseblock 261 at 0x000000414000
[    6.010000] Bad eraseblock 262 at 0x000000418000
[    6.020000] Bad eraseblock 263 at 0x00000041c000
[    6.020000] Bad eraseblock 264 at 0x000000420000
...
...
[   24.280000] Bad eraseblock 4091 at 0x000003fec000
[   24.280000] Bad eraseblock 4092 at 0x000003ff0000
[   24.290000] Bad eraseblock 4093 at 0x000003ff4000
[   24.290000] Bad eraseblock 4094 at 0x000003ff8000
[   24.300000] Bad eraseblock 4095 at 0x000003ffc000
[   24.300000] Creating 3 MTD partitions on "NAND 64MiB 3,3V 8-bit":
[   24.310000] 0x000000000000-0x000000040000 : "booter"
[   24.310000] 0x000000040000-0x000000400000 : "kernel"
[   24.320000] 0x000000400000-0x000004000000 : "rootfs"
[   24.330000] mtd: device 6 (rootfs) set to be root filesystem
[   24.330000] mtdsplit: no squashfs found in "rootfs"
[   24.340000] mtdsplit: no squashfs found in "NAND 64MiB 3,3V 8-bit"
[   24.350000] switch0: Atheros AR8316 rev. 1 switch registered on gpio-ffffffff
[   24.360000] libphy: GPIO Bitbanged MDIO: probed
[   24.370000] switch1: Atheros AR8316 rev. 1 switch registered on ag71xx-mdio.0
[   24.380000] libphy: ag71xx_mdio: probed
[   24.700000] ar8316: Using port 4 as switch port
[   24.740000] ag71xx ag71xx.1: connected to PHY at gpio-ffffffff:00 [uid=004dd041, driver=Atheros AR8216/AR8236/AR8316]
[   24.750000] eth0: Atheros AG71xx at 0xba000000, irq 5, mode:RGMII
[   25.080000] ar8316: Using port 4 as switch port
[   25.110000] ag71xx ag71xx.0: connected to PHY at ag71xx-mdio.0:00 [uid=004dd041, driver=Atheros AR8216/AR8236/AR8316]
[   25.120000] eth1: Atheros AG71xx at 0xb9000000, irq 4, mode:RGMII
[   25.130000] TCP: cubic registered
[   25.130000] NET: Registered protocol family 17
[   25.140000] bridge: automatic filtering via arp/ip/ip6tables has been deprecated. Update your scripts to load br_netfilter if you need this.
[   25.150000] 8021q: 802.1Q VLAN Support v1.8
[   25.150000] rb: no calibration data found
[   25.170000] Freeing unused kernel memory: 2836K (8036b000 - 80630000)
[   25.190000] init: Console is alive
[   25.190000] init: - watchdog -
[   26.200000] init: - preinit -
[   26.350000] random: mktemp urandom read with 41 bits of entropy available
Press the [f] key and hit [enter] to enter failsafe mode
Press the [1], [2], [3] or [4] key and hit [enter] to select the debug level
[   27.740000] eth0: link up (1000Mbps/Full duplex)
[   29.450000] eth0: link down
[   29.460000] procd: - early -
[   29.460000] procd: - watchdog -
[   30.010000] procd: - ubus -
[   31.020000] procd: - init -
Please press Enter to activate this console.
[   32.240000] NET: Registered protocol family 10
[   32.250000] ip6_tables: (C) 2000-2006 Netfilter Core Team
[   32.270000] Loading modules backported from Linux version v4.4-rc5-1913-gc8fdf68
[   32.270000] Backport generated by backports.git backports-20151218-0-g2f58d9d
[   32.280000] ip_tables: (C) 2000-2006 Netfilter Core Team
[   32.300000] nf_conntrack version 0.5.0 (4002 buckets, 16008 max)
[   32.330000] xt_time: kernel timezone is -0000
[   32.400000] PPP generic driver version 2.4.2
[   32.410000] NET: Registered protocol family 24
[   32.440000] PCI: Enabling device 0000:00:13.0 (0000 -> 0002)
[   34.030000] ieee80211 phy0: Atheros AR9280 Rev:2 mem=0xb0000000, irq=41
[   39.790000] device eth0 entered promiscuous mode
[   39.790000] IPv6: ADDRCONF(NETDEV_UP): br-lan: link is not ready
[   39.840000] IPv6: ADDRCONF(NETDEV_UP): eth1: link is not ready
[   39.870000] device eth1.1 entered promiscuous mode
[   39.870000] device eth1 entered promiscuous mode
[   39.910000] IPv6: ADDRCONF(NETDEV_UP): eth1.2: link is not ready
[   41.070000] eth0: link up (1000Mbps/Full duplex)
[   41.070000] eth1: link up (1000Mbps/Full duplex)
[   41.090000] br-lan: port 1(eth0) entered forwarding state
[   41.090000] br-lan: port 1(eth0) entered forwarding state
[   41.100000] IPv6: ADDRCONF(NETDEV_CHANGE): eth1: link becomes ready
[   41.100000] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
[   41.110000] br-lan: port 2(eth1.1) entered forwarding state
[   41.120000] br-lan: port 2(eth1.1) entered forwarding state
[   41.120000] IPv6: ADDRCONF(NETDEV_CHANGE): eth1.2: link becomes ready
[   43.080000] Atheros AR8216/AR8236/AR8316 ag71xx-mdio.0:00: Port 5 is up
[   43.090000] br-lan: port 1(eth0) entered forwarding state
[   43.120000] br-lan: port 2(eth1.1) entered forwarding state
[   80.370000] random: nonblocking pool is initialized

BusyBox v1.23.2 (2017-08-02 15:12:00 BRT) built-in shell (ash)

  _______                     ________        __
 |       |.-----.-----.-----.|  |  |  |.----.|  |_
 |   -   ||  _  |  -__|     ||  |  |  ||   _||   _|
 |_______||   __|_____|__|__||________||__|  |____|
          |__| W I R E L E S S   F R E E D O M
 -----------------------------------------------------
 CHAOS CALMER (Chaos Calmer, r49389)
 -----------------------------------------------------
  * 1 1/2 oz Gin            Shake with a glassful
  * 1/4 oz Triple Sec       of broken ice and pour
  * 3/4 oz Lime Juice       unstrained into a goblet.
  * 1 1/2 oz Orange Juice
  * 1 tsp. Grenadine Syrup
 -----------------------------------------------------

**Comment by AlverGant** [comment]
For the record, I was able to revert to RouterOS with the following process:
Recompile Chaos Calmer with a modified MTD tool that allowed me to forcelly rewrite all NAND blocks including false marked bad blocks (mtd erase -N /dev/mtd6). That alone freed some MBytes just the necessary to do a RouterOS Netinstall.
Once RouterOS got installed it imediately reclaim all NAND storage with 0.1% bad blocks.
Subsequently I was able to finally install OpenWRT through a TFTP and finally wget2nand.
This time without the micro-sd card installed, I had no problems
